### PR TITLE
Static compile darwin binary

### PR DESCRIPTION
**Pull request by notnoop**
Darwin targets don't require CGO, unlike Linux.  Neither do Windows - In fact Windows binaries are statically typed already!

With this change, Nomad follows the rest of HashiCorp products, that statically compile Windows and Darwin targets (and sometimes Linux too!).  Also, enables us to build darwin binaries without needing a macOS machine access.

Nomad relies on cgo for two primary reasons:

1. runc and libcontainer ecosystem for creating and managing linux containers
2. nvidia gpu libraries that link to nvml library

Both are Linux specific and don't apply to Windows or macOS/Darwin.

Some dependencies packages do use cgo (e.g. `net` and `os/user).  These have fallback pure go paths.  I was weary of these pure go implementation, but noticed that Nomad is the only HashiCorp tool that uses CGO - Consul and Vault binary releases are statically linked, and don't differ significantly from Nomad in network capabilities when running on darwin.

I propose we start building static binaries in development as well as next beta/RC releases and pay attention if we see any issues.  We can rollback back change if we see any concerns/issues.

This is an alternative to https://github.com/hashicorp/nomad/pull/7660 .

### CODE REVIEW

File: GNUmakefile [hunk 1]
```diff
@@ -124,7 +130,7 @@ pkg/windows_386/nomad: $(SOURCE_FILES) ## Build Nomad for windows/386
 
 pkg/windows_amd64/nomad: $(SOURCE_FILES) ## Build Nomad for windows/amd64
 	@echo "==> Building $@ with tags $(GO_TAGS)..."
-	@CGO_ENABLED=1 GOOS=windows GOARCH=amd64 \
+	@CGO_ENABLED=$(NON_LINUX_CGO_ENABLED) GOOS=windows GOARCH=amd64 \
```
Comment (notnoop) [inline GNUmakefile:133]: Despite `CGO_ENABLED=1` here, when we cross compiled, go was making a static binary, as I inspected the binary in a Windows machine with `dumpbin`.  I think (but not 100% certain) that we would have needed to set CC and CXX to configure the cross compiler for go to use.

**Comment by notnoop** [comment]
Sadly upon inspecting the source further, I noticed that CPU stats tracking for rawexec driver relies on cgo in [1].  The non-cgo equivalent in [2] returns "not implemented yet" errors.

As such, I declare this option unviable.

[1] https://github.com/hashicorp/nomad/blob/v0.11.0/vendor/github.com/shirou/gopsutil/cpu/cpu_darwin_cgo.go
[2] https://github.com/hashicorp/nomad/blob/v0.11.0/vendor/github.com/shirou/gopsutil/cpu/cpu_darwin_nocgo.go

**Comment by tgross** [comment]
> Neither do Windows - In fact Windows binaries are statically typed already!

I'm surprised at that, and I wonder whether it'll hold up once we solidify the story for Windows 2019 (specifically WSL containers).

**Comment by notnoop** [comment]
Was surprising to me too.  So some caveats: I've been sloppy at what I call static linked here - I mostly meant self-contained binary without requiring C compiler.

Here are my reproduction steps: On a windows machine, I ensured that CGO_ENABLED is enabled, and inspected the dependencies of the binaries:

```
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> $env:CGO_ENABLED=1
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> go env | findstr CGO_
set CGO_ENABLED=1
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> go build -o .\nomad-windows
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> dumpbin /dependents .\nomad-windows
Microsoft (R) COFF/PE Dumper Version 14.25.28614.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file .\nomad-windows

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    kernel32.dll

  Summary

      146000 .data
        1000 .debug_gdb_scripts
        1000 .idata
     2C5D000 .rdata
      F7D000 .symtab
     1EEE000 .text
        1000 .zdebug_abbrev
       CD000 .zdebug_frame
      63D000 .zdebug_info
      31E000 .zdebug_line
      326000 .zdebug_loc
       15000 .zdebug_pubnames
       9A000 .zdebug_pubtypes
       EC000 .zdebug_ranges
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> dumpbin /dependents C:\Users\Mahmood\Downloads\nomad_0.11.0_windows_amd64\nomad.exe
Microsoft (R) COFF/PE Dumper Version 14.25.28614.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file C:\Users\Mahmood\Downloads\nomad_0.11.0_windows_amd64\nomad.exe

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    kernel32.dll

  Summary

      1D9000 .data
        1000 .debug_gdb_scripts
        1000 .idata
     2D67000 .rdata
      F78000 .symtab
     2045000 .text
        1000 .zdebug_abbrev
       CE000 .zdebug_frame
      66E000 .zdebug_info
      346000 .zdebug_line
      34C000 .zdebug_loc
       15000 .zdebug_pubnames
       99000 .zdebug_pubtypes
      102000 .zdebug_ranges
```

Note that `kernel32.dll` is the only dependency.  That's also the case, when I'm compiling with `CGO_ENABLED=0`:

```
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> $env:CGO_ENABLED=0
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> go env | findstr CGO_
set CGO_ENABLED=0
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> go build -o .\nomad-windows-cgo-disabled
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> dumpbin.exe /dependents .\nomad-windows-cgo-disabled
Microsoft (R) COFF/PE Dumper Version 14.25.28614.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file .\nomad-windows-cgo-disabled

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    kernel32.dll

  Summary

      146000 .data
        1000 .debug_gdb_scripts
        1000 .idata
     2C5D000 .rdata
      F7D000 .symtab
     1EEE000 .text
        1000 .zdebug_abbrev
       CD000 .zdebug_frame
      63C000 .zdebug_info
      31E000 .zdebug_line
      326000 .zdebug_loc
       15000 .zdebug_pubnames
       9A000 .zdebug_pubtypes
       E8000 .zdebug_ranges
```

In my Windows machine, if I actually compile an app that requires cgo (found in https://play.golang.org/p/RkdMe-7HbJB ) - I get the following error:

```
PS C:\Users\Mahmood\go\src\github.com\hashicorp\nomad> go build -o windows-cgo .\a\main.go
# command-line-arguments
exec: "gcc": executable file not found in %PATH%
```

When cross compiling in a Linux box, I get the following error:

```
vagrant@linux:~/go/src/github.com/hashicorp/nomad$ CGO_ENABLED=1 GOOS=windows GOARCH=amd64 go build -o windows-cgo ./a/main.go
# runtime/cgo
gcc: error: unrecognized command line option ‘-mthreads’
```

Thus I conclude that Nomad binary, specially when cross compiled from linux, isn't actually using cgo.

FWIW - I suspect we'll find out immediately if Windows starts requiring cgo, as it would fail in the CI  `build_binaries` job - though that's a compile check only; it's possible that we'd run into the case here where a library has a not-implemented non-cgo option.  We should update our windows CI to disable CGO for consistency with release process and to get an early warning.  Will create an issue there.

**Pull request closed without merging**